File: products/LoanHub/apps/frontend/types/analytics.ts
export type AnalyticsMetricFormat = "money" | "integer" | "percent" | "decimal";
export type AnalyticsInsightTone = "positive" | "warning" | "critical" | "neutral";

export type AnalyticsMetric = {
    key: string;
    label: string;
    value: number;
    format: AnalyticsMetricFormat;
    previous_value?: number | null;
    change_percent?: number | null;
    description?: string | null;
};

export type AnalyticsTimePoint = {
    period: string;
    values: Record<string, number>;
};

export type AnalyticsBreakdownPoint = {
    label: string;
    value: number;
    count?: number | null;
    secondary_value?: number | null;
    extra?: Record<string, unknown>;
};

export type AnalyticsScatterPoint = {
    label: string;
    x: number;
    y: number;
    size?: number | null;
    extra?: Record<string, unknown>;
};

export type AnalyticsHeatmapPoint = {
    day: string;
    hour: number;
    value: number;
};

export type AnalyticsInsight = {
    title: string;
    message: string;
    tone: AnalyticsInsightTone;
    action_url?: string | null;
};

export type AnalyticsScope = {
    scope: "platform" | "company" | "borrower";
    company_id?: string | null;
    branch_id?: string | null;
    borrower_id?: string | null;
    role?: string | null;
    date_from: string;
    date_to: string;
    granularity: "day" | "week" | "month";
};

export type AnalyticsDashboard = {
    scope: AnalyticsScope;
    metrics: AnalyticsMetric[];
    series: Record<string, AnalyticsTimePoint[]>;
    breakdowns: Record<string, AnalyticsBreakdownPoint[]>;
    scatter: Record<string, AnalyticsScatterPoint[]>;
    heatmaps: Record<string, AnalyticsHeatmapPoint[]>;
    insights: AnalyticsInsight[];
    permissions: Record<string, boolean>;
    generated_at: string;
};

export type AnalyticsFilters = {
    date_from: string;
    date_to: string;
    granularity?: "day" | "week" | "month";
    branch_id?: string;
};

File: products/LoanHub/apps/backend/database/schemas/analytics.py
from __future__ import annotations

from datetime import date, datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


MetricFormat = Literal["money", "integer", "percent", "decimal"]
InsightTone = Literal["positive", "warning", "critical", "neutral"]


class AnalyticsMetricRead(BaseModel):
    key: str
    label: str
    value: float
    format: MetricFormat = "decimal"
    previous_value: float | None = None
    change_percent: float | None = None
    description: str | None = None


class AnalyticsTimePointRead(BaseModel):
    period: str
    values: dict[str, float] = Field(default_factory=dict)


class AnalyticsBreakdownPointRead(BaseModel):
    label: str
    value: float
    count: int | None = None
    secondary_value: float | None = None
    extra: dict[str, Any] = Field(default_factory=dict)


class AnalyticsScatterPointRead(BaseModel):
    label: str
    x: float
    y: float
    size: float | None = None
    extra: dict[str, Any] = Field(default_factory=dict)


class AnalyticsHeatmapPointRead(BaseModel):
    day: str
    hour: int
    value: int


class AnalyticsInsightRead(BaseModel):
    title: str
    message: str
    tone: InsightTone = "neutral"
    action_url: str | None = None


class AnalyticsScopeRead(BaseModel):
    scope: Literal["platform", "company", "borrower"]
    company_id: str | None = None
    branch_id: str | None = None
    borrower_id: str | None = None
    role: str | None = None
    date_from: date
    date_to: date
    granularity: Literal["day", "week", "month"]


class AnalyticsDashboardRead(BaseModel):
    scope: AnalyticsScopeRead
    metrics: list[AnalyticsMetricRead] = Field(default_factory=list)
    series: dict[str, list[AnalyticsTimePointRead]] = Field(default_factory=dict)
    breakdowns: dict[str, list[AnalyticsBreakdownPointRead]] = Field(default_factory=dict)
    scatter: dict[str, list[AnalyticsScatterPointRead]] = Field(default_factory=dict)
    heatmaps: dict[str, list[AnalyticsHeatmapPointRead]] = Field(default_factory=dict)
    insights: list[AnalyticsInsightRead] = Field(default_factory=list)
    permissions: dict[str, bool] = Field(default_factory=dict)
    generated_at: datetime
